### PR TITLE
Add Price Identifiers

### PR DESCRIPTION
**Pull request by LauLaman**
Thanks for this Implementation. it saves me a lot of work "thinking and inventing the weel"

I've added 2 Identifiers ([8005](https://www.gs1.org/standards/barcodes/application-identifiers/8005?lang=en) and [392x](https://www.gs1.org/standards/barcodes/application-identifiers/3920?lang=en))

I've also put the implemented identifiers in a table in `README.md` and linked them to the GS1.org site

### CODE REVIEW

File: src/Identifiers/PricePerUnitIdentifier.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        return 6;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getFormat(): string
+    {
+        return '/^8005(\d{6})$/';
```
Comment (Dartui) [inline src/Identifiers/PricePerUnitIdentifier.php]: This should be only value format, without identifier

File: README.md [hunk 2]
```diff
@@ -65,18 +65,24 @@ Array
...
+|      17     |      USE BY     | [Expiration date](https://www.gs1.org/standards/barcodes/application-identifiers/17?lang=en)                                                               |
+|      20     |     VARIANT     | [Internal product variant](https://www.gs1.org/standards/barcodes/application-identifiers/20?lang=en)                                                      |
+|      21     |      SERIAL     | [Serial number](https://www.gs1.org/standards/barcodes/application-identifiers/21?lang=en)                                                                 |
+|     422     |      ORIGIN     | [Country of origin of a trade item](https://www.gs1.org/standards/barcodes/application-identifiers/422?lang=en)                                             |
+| 3100 - 3105 | NET WEIGHT (kg) | [Net weight, kilograms](https://www.gs1.org/standards/barcodes/application-identifiers/3100?lang=en)                                                         |
+| 3920 - 3929 |      PRICE      | [Applicable amount payable, single monetary area (variable measure trade item)](https://www.gs1.org/standards/barcodes/application-identifiers/3920?lang=en) |
```
Comment (Dartui) [inline README.md]: Please remove spaces around hyphen in identifier column

File: src/Decoder.php [hunk 3]
```diff
@@ -3,9 +3,11 @@
 
 namespace NGT\Barcode\GS1Decoder;
 
+use App\Grocery\Product\Barcode\Domain\Model\Exception\InvalidBarcodeException;
```
Comment (Dartui) [inline src/Decoder.php]: Remove this line

**Review by Dartui — COMMENTED** [review]
Hi, first of all - thanks for you contribution! Please fix discussions, then it will be ready to merge.

**Comment by LauLaman** [comment]
HI @Dartui

I've applied your requested changes.
I've also setup a test suite and created tests for all implemented Application Identifiers.
~Unfortunately not all Application Identifiers are [implemented correctly](https://github.com/LauLaman/gs1-decoder/runs/1577187689):~
~- Extracting `Batch or lot number (10)` when also a `Serial number (21)` is provided~
~- Extracting `Serial number (21)` when also a `Batch or lot number (10)` is provided~
~- Extracting `Country of origin (422)`~

~For the batch/lot + serial tests i've used the example barcode as found in [GS1us infographic](https://www.gs1us.org/DesktopModules/Bring2mind/DMX/Download.aspx?Command=Core_Download&EntryId=646&language=en-US&PortalId=0&TabId=134) `(01)00614141007349(17)141231(10)A12345B(21)1234` and `(01)00614141999996(10)987654321GFEDCBA`~

~No clue yet  why the country code is not working~


~also `Batch or lot number (10)`  should allow [more characters](https://regexr.com/5iqik)~

**Comment by Dartui** [comment]
@LauLaman you should not use readable by human values which are under barcode. Some identifier has variable length and needs special FNC1 char at the end to indicate that they are ending. This part of code is handling this behaviour:

https://github.com/nextgen-tech/gs1-decoder/blob/595d86aee1996833067c644245507a8d4184bb46/src/Decoder.php#L122-L132

Because of this you can pass `delimiter` parameter to Decoder class and your barcode shoude use it. For example in one of our application we read barcode in JS and translate FNC1 char to `[FNC1]` string.

You can read about this in GS1 specification. Also look at this document at 143 page to know which identifiers might need FNC1 character at the end (every identifier with variable length): https://www.gs1.org/docs/barcodes/GS1_General_Specifications.pdf

**Comment by Dartui** [comment]
@LauLaman you should not use readable by human values which are under barcode. Some identifier has variable length and needs special FNC1 char at the end to indicate that they are ending. This part of code is handling this behaviour:

https://github.com/nextgen-tech/gs1-decoder/blob/595d86aee1996833067c644245507a8d4184bb46/src/Decoder.php#L122-L132

Because of this you can pass `delimiter` parameter to Decoder class and your barcode shoude use it. For example in one of our application we read barcode in JS and translate FNC1 char to `[FNC1]` string.

You can read about this in GS1 specification. Also look at this document at 143 page to know which identifiers might need FNC1 character at the end (every identifier with variable length): https://www.gs1.org/docs/barcodes/GS1_General_Specifications.pdf

**Comment by LauLaman** [comment]
@Dartui gotcha!

I am quite new to the GS1 DataBar world, and have not completed the GS1 General Specifications yet since its quite extensive.
But this has been very helpfull! I've updated the tests to non-human-readable barcodes and included the `delimiter` where needed.


✅. All tests now [succeeds](https://github.com/LauLaman/gs1-decoder/runs/1579256558)

File: phpunit.bootstrap.php [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+declare(strict_types=1);
+require_once('vendor/autoload.php');
+// Add the 'tests/' directory to the include path for unit tests, as we have Test implementation classes that we need to
+// have, but don't want to include in projects that include this library
+$loader = new \Composer\Autoload\ClassLoader();
+$loader->add('NGT\\Barcode\\GS1Decoder\\', ['src/']);
+$loader->add('NGT\\Barcode\\GS1Decoder\\Test\\', ['tests/']);
+$loader->register();
+$loader->setUseIncludePath(true);
```
Comment (Dartui) [inline phpunit.bootstrap.php]: This file is not needed, it can be safely removed. In `phpunit.xml` you can change autoload attribute to `vendor/autoload.php`

File: phpunit.xml.dist [hunk 5]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Dartui) [inline phpunit.xml.dist:1]: While running phpunit I am receiving warning `Your XML configuration validates against a deprecated schema.`. Could you fix that?

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v1
+
+      - name: Install composer dependencies
+        run: composer install --prefer-dist --ignore-platform-reqs
+
+      - name: Run PHPUnit Integration tests
+        run: php7.4 ./vendor/bin/phpunit --testsuite=integration
```
Comment (Dartui) [inline .github/workflows/main.yml]: Could you use [act](https://github.com/nektos/act) to test if it works locally? On my machine I am getting `/github/workflow/1: line 2: composer: command not found`.

You can look at working workflows in other of our projects, for example here:

https://github.com/nextgen-tech/ewus-php/blob/e8820848f99c10fe075ccbe53f0b89ec32966ec4/.github/workflows/main.yml

File: phpunit.xml.dist [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+<phpunit
+        backupGlobals               = "false"
+        backupStaticAttributes      = "false"
+        colors                      = "true"
+        convertErrorsToExceptions   = "true"
+        convertNoticesToExceptions  = "true"
+        convertWarningsToExceptions = "true"
+        processIsolation            = "false"
+        stopOnFailure               = "false"
+        bootstrap                   = "phpunit.bootstrap.php" >
```
Comment (Dartui) [inline phpunit.xml.dist]: This is the line that you need change after removing bootstrap file.

**Comment by LauLaman** [comment]
@Dartui Done!

I've also included phpstan in the test and updated the composer php version to 7.4, since return type `self` is not available in PHP 7.3. (as used in `Identifiers\Abstracts\Identifier::copy()`)

btw i run the actions locally using 
```sh
$ act -P ubuntu-latest=shivammathur/node:latest
```

File: composer.json [hunk 8]
```diff
@@ -9,14 +9,18 @@
         }
     ],
     "require": {
-        "php": "^7.3|^8.0"
+        "php": "^7.4|^8.0"
```
Comment (Dartui) [inline composer.json]: Revert this change and rebase you PR to main - I've changed error with invalid return types few hours ago

File: .phpunit.cache/test-results [hunk 9]
```diff
@@ -0,0 +1 @@
...ate"";d:0.001;s:90:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Due date"";d:0.001;s:98:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Best before date"";d:0.001;s:117:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Net weight, kilograms [no decimals]"";d:0;s:94:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Sell by date"";d:0.001;s:97:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Expiration date"";d:0.001;s:106:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Internal product variant"";d:0;s:105:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Serial number [Numeric]"";d:0;s:110:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Serial number [Alphanumeric]"";d:0;s:115:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Country of origin of a trade item"";d:0;s:109:"NGT\Barcode\GS1Decoder\Test\Integration\DecodeTest::testCanDecode with data set "Date and time of production"";d:0;}}}
```
Comment (Dartui) [inline .phpunit.cache/test-results]: Please remove this

**Review by Dartui — COMMENTED** [review]
I think that this is few last things to fix and this PR will be ready to go!

**Comment by LauLaman** [comment]
@Dartui Done :)
